### ISSUE TITLE
Reader finished with Error RetriesExhaustedException during longevity run

### ISSUE THREAD

**Issue by sumit-bm** [pravega/pravega#3748]
While exercising small workload IO (4 Reader, 3 Writer, scale type as EVENTS_PER_SECOND with retention enabled) often experiencing readers are getting exhausted with `RetriesExhaustedException` after few hours of operation.

Environment details: PKS / K8 with medium cluster:
```
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
```
Pravega version: 0.5.0-2236.5228e2d
Zookeeper Operator : pravega/zookeeper-operator:0.2.1
Pravega Operator: pravega/pravega-operator:0.3.2
```
Experienced first reader exception after ~ 2hours of IO operation:
```
Readers (3/4): events:577,280, events/sec:81, KB/sec:62.51367
New runtime for dfc36f8b-004d-420e-99bf-c061d49b953a  Running for 1h 58m
```
Second reader exception after ~ 6hours of IO operation:
```
Readers (2/4): events:1,177,732, events/sec:53, KB/sec:40.89453
New runtime for dfc36f8b-004d-420e-99bf-c061d49b953a  Running for 6h 8m
```
From Log it seems there are `RetriesExhaustedException` due to `ConnectionFailedException`
```
io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
        at java.util.concurrent.CompletableFuture.reportJoin(Unknown Source)
        at java.util.concurrent.CompletableFuture.join(Unknown Source)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.handleRequest(SegmentInputStreamImpl.java:137)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:122)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.readEvent(EventSegmentReaderImpl.java:75)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.read(EventSegmentReaderImpl.java:62)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:117)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:90)
        at io.pravega.longevity.testworker.workers.readers.PravegaReader.start(PravegaReader.java:99)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.failAllInflight(AsyncSegmentInputStreamImpl.java:265)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.closeConnection(AsyncSegmentInputStreamImpl.java:237)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.access$100(AsyncSegmentInputStreamImpl.java:42)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl$ResponseProcessor.connectionDropped(AsyncSegmentInputStreamImpl.java:65)
        at io.pravega.client.netty.impl.FlowHandler.lambda$channelUnregistered$0(FlowHandler.java:174)
        at java.util.concurrent.ConcurrentHashMap.forEach(Unknown Source)
        at io.pravega.client.netty.impl.FlowHandler.channelUnregistered(FlowHandler.java:173)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelUnregistered(DefaultChannelPipeline.java:1412)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.DefaultChannelPipeline.fireChannelUnregistered(DefaultChannelPipeline.java:865)
        at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:830)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:326)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        ... 1 more
Caused by: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
        at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
        ... 35 more
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException
        ... 30 more
```
And `ConnectionFailedException: null` on `WireCommands.ReadSegment` with `READ_SEGMENT` `segment=longevity/smallScale/91.#epoch.53` yields `readers.PravegaReader - Reader finished with Error` 
```
2019-05-02 16:53:24,811 7075568 [epollEventLoopGroup-4-7] INFO  i.p.client.netty.impl.FlowHandler - Closing Flow 519 for endpoint 172.24.66.16
2019-05-02 16:53:24,811 7075568 [epollEventLoopGroup-4-7] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Connection failed due to a io.pravega.shared.protocol.netty.ConnectionFailedException. Read requests will be retransmitted.
2019-05-02 16:53:24,811 7075568 [epollEventLoopGroup-4-7] DEBUG i.p.c.s.i.AsyncSegmentInputStreamImpl - ConnectionFailedException observed when sending request WireCommands.ReadSegment(type=READ_SEGMENT, segment=longevity/smallScale/91.#epoch.53, offset=1029039, suggestedLength=1047779, delegationToken=, requestId=2229088026624)
io.pravega.shared.protocol.netty.ConnectionFailedException: null

2019-05-02 16:53:24,967 7075724 [pool-7-thread-4] ERROR i.p.l.t.w.readers.PravegaReader - Reader finished with Error
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
        at java.util.concurrent.CompletableFuture.reportJoin(Unknown Source)
        at java.util.concurrent.CompletableFuture.join(Unknown Source)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.handleRequest(SegmentInputStreamImpl.java:137)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:122)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.readEvent(EventSegmentReaderImpl.java:75)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.read(EventSegmentReaderImpl.java:62)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:117)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:90)
        at io.pravega.longevity.testworker.workers.readers.PravegaReader.start(PravegaReader.java:99)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.failAllInflight(AsyncSegmentInputStreamImpl.java:265)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.closeConnection(AsyncSegmentInputStreamImpl.java:237)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.access$100(AsyncSegmentInputStreamImpl.java:42)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl$ResponseProcessor.connectionDropped(AsyncSegmentInputStreamImpl.java:65)
        at io.pravega.client.netty.impl.FlowHandler.lambda$channelUnregistered$0(FlowHandler.java:174)
        at java.util.concurrent.ConcurrentHashMap.forEach(Unknown Source)
        at io.pravega.client.netty.impl.FlowHandler.channelUnregistered(FlowHandler.java:173)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelUnregistered(DefaultChannelPipeline.java:1412)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.DefaultChannelPipeline.fireChannelUnregistered(DefaultChannelPipeline.java:865)
        at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:830)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:326)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        ... 1 common frames omitted
Caused by: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
        at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
        ... 35 common frames omitted
```
At the same time of issue with readers segmentstore logs showing following with `segment=longevity/smallScale/91.#epoch.53`
```
2019-05-02 16:53:24,809 7416144 [storage-io-49] ERROR i.p.s.s.h.h.PravegaRequestProcessor - [requestId=2229088026624] Error (Segment = 'longevity/smallScale/91.#epoch.53', Operation = 'readSegment')
java.io.IOException: Bad file descriptor
        at sun.nio.ch.FileDispatcherImpl.pread0(Native Method)
        at sun.nio.ch.FileDispatcherImpl.pread(FileDispatcherImpl.java:52)
        at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:220)
        at sun.nio.ch.IOUtil.read(IOUtil.java:197)
        at sun.nio.ch.FileChannelImpl.readInternal(FileChannelImpl.java:741)
        at sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:727)
        at io.pravega.storage.filesystem.FileSystemStorage.doRead(FileSystemStorage.java:245)
        at io.pravega.storage.filesystem.FileSystemStorage.lambda$read$1(FileSystemStorage.java:129)
        at io.pravega.storage.filesystem.FileSystemStorage.execute(FileSystemStorage.java:405)
        at io.pravega.storage.filesystem.FileSystemStorage.read(FileSystemStorage.java:129)
        at io.pravega.segmentstore.storage.rolling.RollingStorage.read(RollingStorage.java:179)
        at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$read$8(AsyncStorageWrapper.java:128)
        at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$execute$13(AsyncStorageWrapper.java:184)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)



2019-05-02 16:53:24,810 7416145 [storage-io-49] ERROR i.p.s.s.h.h.PravegaRequestProcessor - [requestId=2229088026624] Error (Segment = 'longevity/smallScale/91.#epoch.53', Operation = 'readSegment')
java.lang.IllegalStateException: Unknown exception.
        at io.pravega.segmentstore.server.host.handler.PravegaRequestProcessor.handleException(PravegaRequestProcessor.java:964)
        at io.pravega.segmentstore.server.host.handler.PravegaRequestProcessor.lambda$handleReadResult$5(PravegaRequestProcessor.java:268)
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:870)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:852)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.pravega.segmentstore.server.reading.ReadResultEntryBase.fail(ReadResultEntryBase.java:121)
        at io.pravega.common.function.Callbacks.invokeSafely(Callbacks.java:54)
        at io.pravega.common.concurrent.Futures.lambda$exceptionListener$2(Futures.java:240)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.pravega.segmentstore.server.reading.StorageReadManager$Request.fail(StorageReadManager.java:403)
        at io.pravega.segmentstore.server.reading.StorageReadManager$Request.access$200(StorageReadManager.java:247)
        at io.pravega.segmentstore.server.reading.StorageReadManager.lambda$executeStorageRead$3(StorageReadManager.java:140)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.pravega.common.function.Callbacks.invokeSafely(Callbacks.java:54)
        at io.pravega.common.concurrent.Futures.lambda$exceptionListener$2(Futures.java:240)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1595)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Bad file descriptor
        at sun.nio.ch.FileDispatcherImpl.pread0(Native Method)
        at sun.nio.ch.FileDispatcherImpl.pread(FileDispatcherImpl.java:52)
        at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:220)
        at sun.nio.ch.IOUtil.read(IOUtil.java:197)
        at sun.nio.ch.FileChannelImpl.readInternal(FileChannelImpl.java:741)
        at sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:727)
        at io.pravega.storage.filesystem.FileSystemStorage.doRead(FileSystemStorage.java:245)
        at io.pravega.storage.filesystem.FileSystemStorage.lambda$read$1(FileSystemStorage.java:129)
        at io.pravega.storage.filesystem.FileSystemStorage.execute(FileSystemStorage.java:405)
        at io.pravega.storage.filesystem.FileSystemStorage.read(FileSystemStorage.java:129)
        at io.pravega.segmentstore.storage.rolling.RollingStorage.read(RollingStorage.java:179)
        at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$read$8(AsyncStorageWrapper.java:128)
        at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$execute$13(AsyncStorageWrapper.java:184)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
        ... 7 common frames omitted
```

**Comment by fpj** [comment]
@sumit-bm do you know whether this is related to #3646?

**Comment by sumit-bm** [comment]
@fpj there was no external IO going on in #3646 and one exception was similar `Bad file descriptor`, could come from read on `_system` scope over there.

Also in this issue reader handles are failing, after `Bad file descriptor` there are many other exception `PravegaRequestProcessor - [requestId=2229088026624] Error (Segment = 'longevity/smallScale/91.#epoch.53', Operation = 'readSegment')
java.lang.IllegalStateException: Unknown exception.` which we haven't seen on #3646.

**Comment by sachin-j-joshi** [comment]
@sumit-bm I am trying to understand the source of bad descriptor exception in our setup. 
Bad file descriptor or stale descriptor exceptions are generally result of underlying nfs persistent volume being re-initialized. (This should not happen in normal circumstances).

I am not seeing similar exceptions in the system tests we run everyday multiple times against NFS/Isilon  
I would like to understand the setup for longevity tests so I can create a smaller repro for this issue.

Also - we can not completely rule out possibility of concurrent read and delete. But looking at the scenario described here I can not tell if the segment was deleted while readers were waiting.

**Comment by sumit-bm** [comment]
@sachin-j-joshi I am not just talking over here the `Bad file descriptor` issue, overall this issue is much severe as Pravega read handles are failing. Please go through the logs I posted in internal channel and let us know why Readers are failing and why there are retries exhausted?

**Comment by sachin-j-joshi** [comment]
*Quick update* - A quick look at the logs reveal that most of the StorageWriter errors are ONLY in pravega-pravega-segmentstore-2 ,  not in segment store instance pravega-pravega-segmentstore-0 or pravega-pravega-segmentstore-1.  More interestingly none of the I/O operations succeed for that pod. Before the bad descriptor error happens , lots of other writes fail.

The stack trace and the error message is the most unhelpful message I have seen so far

>  2019-05-02 15:00:35,282 646617 [storage-io-69] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[19]: Iteration[3140].Error.
> java.io.IOException: I/O error
> 	at sun.nio.ch.FileDispatcherImpl.force0(Native Method)
> 	at sun.nio.ch.FileDispatcherImpl.force(FileDispatcherImpl.java:76)
> 	at sun.nio.ch.FileChannelImpl.force(FileChannelImpl.java:388)
> 	at io.pravega.storage.filesystem.FileSystemStorage.doWrite(FileSystemStorage.java:323)
> 	at io.pravega.storage.filesystem.FileSystemStorage.lambda$write$6(FileSystemStorage.java:155)
> 	at io.pravega.storage.filesystem.FileSystemStorage.execute(FileSystemStorage.java:405)
> 	at io.pravega.storage.filesystem.FileSystemStorage.write(FileSystemStorage.java:155)
> 	at io.pravega.segmentstore.storage.rolling.RollingStorage.write(RollingStorage.java:324)
> 	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$write$2(AsyncStorageWrapper.java:93)
> 	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$runAsync$12(AsyncStorageWrapper.java:171)
> 	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$execute$13(AsyncStorageWrapper.java:184)
> 	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
> 

> 
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)

**Comment by sachin-j-joshi** [comment]
The root cause seems to be following 

1.  Session with zookkepper times out

> 2019-05-02 14:49:49,141 476  [main] WARN  o.a.curator.CuratorZookeeperClient - session timeout [10000] is less than connection timeout [15000]

2. which results in *no segment assignment found* 

> 2019-05-02 14:49:49,647 982  [core-2] WARN  i.p.s.s.h.ZKSegmentContainerMonitor - No segment container assignments found

3.  However there are several incoming requests which result in *wrong host* reply

> WARN  i.p.s.s.h.h.PravegaRequestProcessor - [requestId=296] Wrong host. Segment = '_system/_tables/_system/streamsInScope.#.6259c5ad-0505-4a0c-a57b-d833760e69ac' (Container 4) is not owned. Operation = 'readTable').

> 2019-05-02 14:55:36,697 348032 [core-24] WARN  i.p.s.s.h.h.PravegaRequestProcessor - [requestId=339] Segment '_system/_tables/longevity/_RGLongevitySmallScale/metadata.#.1582d9cc-0573-446c-a72c-1a0aaf449896' does not exist and cannot perform operation 'readTable'.

4. later on

> 2019-05-02 14:50:01,844 13179 [core-7] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[19-1]: SegmentMetadata has a StorageLength (0) that is different than the actual one (189) - updating metadata.
> 2019-05-02 14:50:01,844 13179 [core-7] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[19-1]: Initialized. StorageLength = 189, Sealed = false.
> 2019-05-02 14:50:01,844 13179 [core-7] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[19-1]: State changed from NotInitialized to Writing.
> 2019-05-02 14:50:01,867 13202 [core-12] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[0-4]: SegmentMetadata has a StorageLength (0) that is different than the actual one (18) - updating metadata.
> 

5. There are appends 

> 2019-05-02 15:00:04,190 615525 [epollEventLoopGroup-11-4] INFO  i.p.s.s.h.handler.AppendProcessor - Received hello from connection: [id: 0xe31e09d7, L:/172.24.66.16:12345 - R:/172.24.43.5:43916]
> 2019-05-02 15:00:04,190 615525 [epollEventLoopGroup-11-4] INFO  i.p.s.s.h.handler.AppendProcessor - Setting up appends for writer: 0dfab52b-c616-46c5-bdab-78550987a091 on segment: longevity/_RGLongevitySmallTx/0.#epoch.0
> 2019-05-02 15:00:04,258 615593 [epollEventLoopGroup-11-2] INFO  i.p.s.s.h.handler.AppendProcessor - Setting up appends for writer: 0ae9864a-c6f0-40af-9578-21a47d340fa2 on segment: longevity/_RGLongevitySmallTx/0.#epoch.0
> 

6. Storage Writer fails

> 2019-05-02 15:00:35,282 646617 [storage-io-69] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[19]: Iteration[3140].Error.
> java.io.IOException: I/O error
> 	at sun.nio.ch.FileDispatcherImpl.force0(Native Method)
> 	at sun.nio.ch.FileDispatcherImpl.force(FileDispatcherImpl.java:76)
> 	at sun.nio.ch.FileChannelImpl.force(FileChannelImpl.java:388)
> 	at io.pravega.storage.filesystem.FileSystemStorage.doWrite(FileSystemStorage.java:323)
> 	at io.pravega.storage.filesystem.FileSystemStorage.lambda$write$6(FileSystemStorage.java:155)
> 	at io.pravega.storage.filesystem.FileSystemStorage.execute(FileSystemStorage.java:405)
> 	at io.pravega.storage.filesystem.FileSystemStorage.write(FileSystemStorage.java:155)
> 	at io.pravega.segmentstore.storage.rolling.RollingStorage.write(RollingStorage.java:324)
> 	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$write$2(AsyncStorageWrapper.java:93)
> 	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$runAsync$12(AsyncStorageWrapper.java:171)
> 	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$execute$13(AsyncStorageWrapper.java:184)
> 	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)

7. There is resulting bad offset

> 2019-05-02 15:22:47,892 1979227 [storage-io-177] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[16]: Iteration[8463].Error.
> io.pravega.segmentstore.contracts.BadOffsetException: [Segment 'longevity/smallScale/5.#epoch.3$offset.0'] Bad Offset. Expected 4034658, given 1318643.
> 	at io.pravega.storage.filesystem.FileSystemStorage.doWrite(FileSystemStorage.java:310)
> 	at io.pravega.storage.filesystem.FileSystemStorage.lambda$write$6(FileSystemStorage.java:155)
> 	at io.pravega.storage.filesystem.FileSystemStorage.execute(FileSystemStorage.java:405)
> 	at io.pravega.storage.filesystem.FileSystemStorage.write(FileSystemStorage.java:155)
> 	at io.pravega.segmentstore.storage.rolling.RollingStorage.write(RollingStorage.java:324)

**NOW the critical error occurs in StorageWriter**

> 2019-05-02 15:22:50,263 1981598 [core-19] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[16]: Iteration[8464].CriticalError.
> io.pravega.segmentstore.server.writer.ReconciliationFailureException: Unable to reconcile operation 'AggregatedAppend: SegmentId = 89, Offsets = [1318643-4034658), SeqNo = 987' because of data differences at SegmentOffset 1318649. Segment = longevity/smallScale/5.#epoch.3, Storage: Length=4034658(Not Sealed), Metadata: Length=1318643(Not Sealed)
> 	at io.pravega.segmentstore.server.writer.SegmentAggregator.verifySame(SegmentAggregator.java:1391)
> 	at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$reconcileData$54(SegmentAggregator.java:1381)
> 	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
> 	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
> 	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)

**It shuts down the containers**

> 2019-05-02 15:22:50,264 1981599 [core-19] INFO  i.p.c.c.AbstractThreadPoolService - StorageWriter[16]: Stopping.
> 2019-05-02 15:22:50,267 1981602 [core-19] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[16-1]: State changed from Writing to Closed.
> 2019-05-02 15:22:50,267 1981602 [core-19] INFO  i.p.s.s.tables.WriterTableProcessor - TableProcessor[16-1]: Closed.
> 2019-05-02 15:22:50,267 1981602 [core-19] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[16-22]: State changed from Writing to Closed.
> 2019-05-02 15:22:50,267 1981602 [core-19] INFO  i.p.s.s.tables.WriterTableProcessor - TableProcessor[16-22]: Closed.
> 2019-05-02 15:22:50,267 1981602 [core-19] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[16-9]: State changed from Writing to Closed.
> 2019-05-02 15:22:50,267 1981602 [core-19] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[16-89]: State changed from Reconciling to Closed.
> 2019-05-02 15:22:50,268 1981603 [core-19] INFO  i.p.c.c.AbstractThreadPoolService - StorageWriter[16]: Stopped.
> 2019-05-02 15:22:50,269 1981604 [core-7] WARN  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[16]: Writer failed. Shutting down StreamSegmentContainer.

**Comment by sachin-j-joshi** [comment]
At this point not sure if writer fails because of earlier multiple metadata updates and container reassignments or the io failure is independent of that. 

The reason for container shutdown is this

> 2019-05-02 15:22:50,269 1981604 [core-7] WARN  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[16]: Writer failed. Shutting down StreamSegmentContainer.
> java.util.concurrent.CompletionException: io.pravega.segmentstore.server.writer.ReconciliationFailureException: Unable to reconcile operation 'AggregatedAppend: SegmentId = 89, Offsets = [1318643-4034658), SeqNo = 987' because of data differences at SegmentOffset 1318649. Segment = longevity/smallScale/5.#epoch.3, Storage: Length=4034658(Not Sealed), Metadata: Length=1318643(Not Sealed)
> 	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
> 	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
> 	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
> 	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
> 	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> Caused by: io.pravega.segmentstore.server.writer.ReconciliationFailureException: Unable to reconcile operation 'AggregatedAppend: SegmentId = 89, Offsets = [1318643-4034658), SeqNo = 987' because of data differences at SegmentOffset 1318649. Segment = longevity/smallScale/5.#epoch.3, Storage: Length=4034658(Not Sealed), Metadata: Length=1318643(Not Sealed)
> 	at io.pravega.segmentstore.server.writer.SegmentAggregator.verifySame(SegmentAggregator.java:1391)
> 	at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$reconcileData$54(SegmentAggregator.java:1381)
> 	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)

**Comment by sachin-j-joshi** [comment]
See also https://github.com/pravega/pravega/issues/3747

**Comment by sachin-j-joshi** [comment]
There is only ONE container restart and then the container is successfully restarted

> 2019-05-02 15:22:59,754 1991089 [core-27] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 1, 17, 2, 18, 3, 19, 4, 20], Current = [0, 1, 17, 2, 18, 3, 19, 4, 20], PendingTasks = [], ToStart = [16], ToStop = [].
> 2019-05-02 15:22:59,754 1991089 [core-27] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Starting Container 16.
> 2019-05-02 15:22:59,754 1991089 [core-27] INFO  i.p.s.s.impl.rocksdb.RocksDBCache - RocksDBCache[Container_16]: Created empty database directory '/tmp/pravega/cache/cache_Container_16'.
> 2019-05-02 15:22:59,789 1991124 [core-27] INFO  i.p.s.s.impl.rocksdb.RocksDBCache - RocksDBCache[Container_16]: Initialized.
> 2019-05-02 15:22:59,790 1991125 [core-27] INFO  i.p.s.s.impl.rocksdb.RocksDBCache - RocksDBCache[Container_16_TableKeys]: Created empty database directory '/tmp/pravega/cache/cache_Container_16_TableKeys'.
> 2019-05-02 15:22:59,830 1991165 [core-27] INFO  i.p.s.s.impl.rocksdb.RocksDBCache - RocksDBCache[Container_16_TableKeys]: Initialized.
> 2019-05-02 15:22:59,830 1991165 [core-27] INFO  i.p.s.server.CacheManager - CacheManager Registered io.pravega.segmentstore.server.tables.ContainerKeyCache@2e75a450.
> 2019-05-02 15:22:59,831 1991166 [core-27] INFO  i.p.s.s.s.StreamSegmentContainerRegistry - Registered SegmentContainer 16.
> 2019-05-02 15:22:59,831 1991166 [core-27] INFO  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[16]: Starting.
> 2019-05-02 15:22:59,831 1991166 [core-27] INFO  i.p.s.server.logs.DurableLog - DurableLog[16]: Starting.
> 2019-05-02 15:22:59,834 1991169 [core-27] INFO  i.p.s.s.impl.bookkeeper.Ledgers - Log[16]: Fenced out Ledger Id = 77, Sequence = 2, Status = Unknown.
> 2019-05-02 15:22:59,839 1991174 [main-EventThread] INFO  o.a.b.client.LedgerCreateOp - Ensemble: [pravega-bookie-3.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181] for ledger: 878
> 2019-05-02 15:22:59,839 1991174 [core-27] INFO  i.p.s.s.i.bookkeeper.BookKeeperLog - Log[16]: Created Ledger 878.
> 2019-05-02 15:22:59,848 1991183 [core-27] INFO  i.p.s.s.i.bookkeeper.BookKeeperLog - Log[16] Metadata persisted (Version = 4, Epoch = 3, LedgerCount = 2, Truncate = (77-599)).
> 2019-05-02 15:22:59,848 1991183 [core-27] INFO  i.p.s.s.i.bookkeeper.BookKeeperLog - Log[16] Metadata updated (Version = 4, Epoch = 3, LedgerCount = 2, Truncate = (77-599)).
> 2019-05-02 15:22:59,848 1991183 [core-27] INFO  i.p.s.s.i.bookkeeper.BookKeeperLog - Log[16]: Initialized (Epoch = 3, UpdateVersion = 4).
> 2019-05-02 15:22:59,848 1991183 [core-27] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[16] Recovery started.
> 2019-05-02 15:22:59,848 1991183 [core-27] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[16]: Enter RecoveryMode.
> 2019-05-02 15:22:59,848 1991183 [core-27] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[16]: Reset.
> 2019-05-02 15:22:59,848 1991183 [core-27] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[16] Enter RecoveryMode.
> 2019-05-02 15:22:59,848 1991183 [core-27] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[16]: Cleared.
> 2019-05-02 15:23:00,373 1991708 [core-27] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[16]: Starting recovery from Sequence Number 692 (skipped 1 Operations and 0 Data Frames).
> 2019-05-02 15:23:00,373 1991708 [core-27] INFO  i.p.s.s.l.ContainerMetadataUpdateTransaction - MetadataUpdate[16-0]: Recovering MetadataCheckpointOperation with SequenceNumber 692.
> 2019-05-02 15:23:00,374 1991709 [core-27] INFO  i.p.s.server.CacheManager - CacheManager Registered ReadIndex[16-89] (longevity/smallScale/5.#epoch.3).
> 2019-05-02 15:23:00,374 1991709 [core-27] INFO  i.p.s.server.CacheManager - CacheManager Registered ReadIndex[16-9] (longevity/_RGLongevitySmallTx/0.#epoch.0).
> 2019-05-02 15:23:00,428 1991763 [core-27] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[16]: Reset.
> 2019-05-02 15:23:00,428 1991763 [core-27] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[16]: MapStreamSegment SegmentId = 1, Name = '_system/containers/metadata_16', Active = 1
> 2019-05-02 15:23:00,428 1991763 [core-27] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[16]: MapStreamSegment SegmentId = 22, Name = '_system/_tables/completedTransactionsBatches', Active = 2
> 2019-05-02 15:23:00,429 1991764 [core-27] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[16]: MapStreamSegment SegmentId = 9, Name = 'longevity/_RGLongevitySmallTx/0.#epoch.0', Active = 3
> 2019-05-02 15:23:00,429 1991764 [core-27] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[16]: MapStreamSegment SegmentId = 89, Name = 'longevity/smallScale/5.#epoch.3', Active = 4
> 2019-05-02 15:23:00,429 1991764 [core-27] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[16] Recovery completed. Epoch = 3, Items Recovered = 2109, Time = 581ms.
> 2019-05-02 15:23:00,429 1991764 [core-27] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[16]: Exit RecoveryMode.
> 2019-05-02 15:23:00,429 1991764 [core-27] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[16] Exit RecoveryMode.
> 2019-05-02 15:23:00,429 1991764 [core-27] INFO  i.p.c.c.AbstractThreadPoolService - OperationProcessor[16]: Started.
> 2019-05-02 15:23:00,429 1991764 [core-4] INFO  i.p.s.server.logs.DurableLog - DurableLog[16]: Started (Online).

**Comment by sachin-j-joshi** [comment]
@shiveshr @andreipaduroiu Based on my reading of the logs and understanding of the code, this issue is not **directly** caused by the tier-2 , 
**Some observations** 
1.  There are 3 segment stores , only one segmentstore-2 fails to write anything to the tier-2.
2. The segmentstore-2 starts at 2019-05-02 **14:49:48** and encounters first IO error immediately around 2019-05-02 **15:00:35**,282.  
3.  The container encounters critical error around 2019-05-02 **15:22:50**,263 1981598 and restarts successfully around 2019-05-02 **15:23:00**,429 
4. The reader is actually getting errors almost after 2 hours  ~ around 2019-05-02 **16:53:24,967** due to **Connection errors**.  

5. Before this time we see several IO errors regularly so in sense there is nothing specifically significant about the IO exception encountered during the same time in segment store.

My initial theory was that container restart is causing the problem by dropping futures waiting on the reads. But there is about ~2 hrs gap between the container restart and when reads start getting failures because of connection errors. **More importantly several reads succeed many times during that timeframe** 

At this point I think StorageWriter failures are not directly causing the errors with readers.

CC: @sumit-bm

**Comment by sachin-j-joshi** [comment]
This particular issue does not seem to be related to tier-2 failures. @shrids 
 is investigating related to that issue https://github.com/pravega/pravega/issues/3754 which might be the root cause.

**Comment by andreipaduroiu** [comment]
This is the same as #3646. They should be merged into one.
This may or may not be related to #3647. #3647 records a general `IO Error` when _writing_ to a file using NFS, while this issue records a `Bad File Descriptor` when _reading_ from a file in NFS. They both point to underlying Tier 2 problems.

All of our Tier 2 reads and Writes open new `FileChannel`s so there is nothing cached on our end, unless `FileChannel.open` caches handles internally.

**Comment by sachin-j-joshi** [comment]
To be correct the IO related issue is captured here https://github.com/pravega/pravega/issues/3747

This particular issue is about the stalled readers and we do not think #3747 is the root cause for _this issue._

**Comment by Raghuveera-Rao** [comment]
Readers dropped to 2/3 and then 1/3 and eventually 0/3 .
Longevity i/o small-scale-young (has 3 writers and 3 readers) was run. I/O ran for about 2h 52 m (approx 3 hour) before one of the readers went down.
Seeing **RetriesExhaustedException** and **java.lang.IllegalStateException: Segment already exists** in longevity manager log just before the reader went to 2/3

```
INFO  [2019-05-10 13:49:19,840] io.pravega.longevity.testmanager.manager.TestPerformanceCollector: Received Error From task taskID : io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Segment already exists: longevity/smallScaleYoung/57.#epoch.30
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Segment already exists: longevity/smallScaleYoung/57.#epoch.30
        at java.util.concurrent.CompletableFuture.reportJoin(Unknown Source)
        at java.util.concurrent.CompletableFuture.join(Unknown Source)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.handleRequest(SegmentInputStreamImpl.java:137)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:122)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.readEvent(EventSegmentReaderImpl.java:75)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.read(EventSegmentReaderImpl.java:62)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:117)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:90)
        at io.pravega.longevity.testworker.workers.readers.PravegaReader.start(PravegaReader.java:99)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Segment already exists: longevity/smallScaleYoung/57.#epoch.30
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.failAllInflight(AsyncSegmentInputStreamImpl.java:265)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.closeConnection(AsyncSegmentInputStreamImpl.java:237)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.access$100(AsyncSegmentInputStreamImpl.java:42)
```
Seeing the **IllegalStateException: Segment already exists** in controller logs

```
2019-05-10 13:56:43,369 12310146 [epollEventLoopGroup-2-2] ERROR i.p.controller.server.SegmentHelper - getSegmentInfo longevity/smallScaleYoung/57.#epoch.30 failed
java.lang.IllegalStateException: Segment already exists: longevity/smallScaleYoung/57.#epoch.30
        at io.pravega.shared.protocol.netty.FailingReplyProcessor.segmentAlreadyExists(FailingReplyProcessor.java:75)
        at io.pravega.controller.server.SegmentStoreConnectionManager$ReusableReplyProcessor.execute(SegmentStoreConnectionManager.java:235)
        at io.pravega.controller.server.SegmentStoreConnectionManager$ReusableReplyProcessor.segmentAlreadyExists(SegmentStoreConnectionManager.java:258)
        at io.pravega.shared.protocol.netty.WireCommands$SegmentAlreadyExists.process(WireCommands.java:238)
        at io.pravega.shared.protocol.netty.ReplyProcessor.process(ReplyProcessor.java:20)
        at io.pravega.client.netty.impl.FlowHandler.lambda$channelRead$2(FlowHandler.java:196)
        at java.util.Optional.ifPresent(Optional.java:159)

```

Complete k8s logs we will share in slack.

**Comment by sachin-j-joshi** [comment]
@shiveshr @shrids This issue is looking more like controller related issue.

**Comment by fpj** [comment]
The discussion in this issue is a bit confusing. The high-level incorrect behavior is that readers are giving up (retries exhausted). We have then observed bad file descriptors, and in the last few comments errors in the controller.

Do I understand it right that there is no clear understanding as of why the reader is giving up? To clarify, the reader giving up isn't an incorrect behavior, it is expected during periods of instability, which are not necessarily related to Pravega. It is only an issue in the case that the Pravega cluster is in a bad state, but there is no clear evidence in the comments that this is the case. For example, it is not clear that the error on the segment existing is causing the readers to die.

**Comment by sachin-j-joshi** [comment]
1. There is no direct chain of events in StorageWriter that seems to have _caused_ the readers to stall. 
There is one container restart because of data mismatch. But the system recovers. Writers and readers make progress for next **two hours** . From code perspective readers are not dependent on writer succeeding.
2. The readers are stalled most likely because of combination of following errors. 

- https://github.com/pravega/pravega/issues/3754
- https://github.com/pravega/pravega/issues/3624
- https://github.com/pravega/pravega/issues/3770

3. On a separate front, we identified the setup issue with isilon that we were using. (it is being reconfigured.)

**Comment by sachin-j-joshi** [comment]
1. This issue is caused by underlying nfs mount for segmentstore-2 throwing "Bad File Descriptor" error while _reading the segments_ . The root cause seems to be the hardware/configuration issue we recently found with the isilon used for testing .

2.  During read operation when we encounter an unknown error _we close the client connection._ This causes the client to retry the operation with exponential backoff. 
https://github.com/pravega/pravega/blob/master/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java#L963

The interaction between client and segment store seems to be correct. It forces Client to retry.

3. The underlying problem persists and none of the retries succeed. They all fail with same exception
The reader first encounters the error at **2019-05-02 16:49:48,286** and retries several times before retries are exhausted. 

The error that we are encountering ( "Bad File Descriptor") is coming from the underlying nfs driver/device. The error persists and as such otherwise non-recoverable and critical.
Client is forced to retry a number of times and reports error when it has retried sufficient number of times.

**Also note -** 
same underlying nfs issue is affecting _write_ path, the writer failures (and subsequent reconciliation failures, segment container restarts etc) are **not** causing this issue. That is a separate issue tracked with https://github.com/pravega/pravega/issues/3747. 
The StorageWriter exceptions on write path issue need to be tracked separately as they are not directly causing **this** issue - although they have same underlying root cause.

**There are other client/controller related issues** that are not caused by this issue and are being actively investigated/worked on.
#3754
#3624
#3770

**Comment by sachin-j-joshi** [comment]
This issue is caused by hardware that is not in healthy state. 
At this point there is nothing actionable to be done at the code level.
CC: @sumit-bm

**Comment by sumit-bm** [comment]
Will retest same longevity concurrent workloads and evaluate.

**Comment by sumit-bm** [comment]
Tested with build #0.5.0-2260.b477eec and longevity working fine with 5d+

Small IO workload:
```
INFO  [2019-06-10 11:21:20,439] io.pravega.longevity.utils.PerformanceUtils: Readers (4/4): events:45,356,732, events/sec:48, KB/sec:37.461914
INFO  [2019-06-10 11:21:20,439] io.pravega.longevity.utils.PerformanceUtils: New runtime for 77b9ced6-ce69-4580-958e-dfabf63c631c  Running for 5d 17h
INFO  [2019-06-10 11:21:20,439] io.pravega.longevity.utils.PerformanceUtils: Dumping stats for stream longevity/smallScale
INFO  [2019-06-10 11:21:20,439] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:37,702,821, events/sec:40, KB/sec:31.139648

- 
```
Transactions:
```
INFO  [2019-06-10 11:20:57,539] io.pravega.longevity.utils.PerformanceUtils: Readers (1/1): events:15,076,034, events/sec:16, KB/sec:12.451172
INFO  [2019-06-10 11:20:57,539] io.pravega.longevity.utils.PerformanceUtils: New runtime for 20a927bb-96c2-4372-b9f3-b91df6ec41ab  Running for  5d 17h
INFO  [2019-06-10 11:20:57,539] io.pravega.longevity.utils.PerformanceUtils: Dumping stats for stream longevity/smallTx
INFO  [2019-06-10 11:20:57,539] io.pravega.longevity.utils.PerformanceUtils: Writers (1/1): events:15,076,024, events/sec:16, KB/sec:12.451172
INFO  [2019-06-10 11:20:57,618] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 631260, perSec:0.6407664233576642, committed:316026, failed:0, aborted:315233, aborted but read: 0

```

**Comment by sachin-j-joshi** [comment]
Thanks @sumit-bm . 
I think it is okay to close this issue. Please reopen if repros again.
